File: react-native-totvs-clockin-vision/android/src/main/java/com/totvs/clockin/vision/face/FaceTransformer.kt
package com.totvs.clockin.vision.face

import android.graphics.PointF
import android.graphics.RectF
import android.util.Size
import com.totvs.camera.view.GraphicOverlay
import com.totvs.camera.vision.face.FaceObject
import com.totvs.camera.vision.face.Landmark
import com.totvs.camera.vision.face.LeftEye
import com.totvs.camera.vision.face.Nose
import com.totvs.camera.vision.stream.VisionReceiver
import com.totvs.clockin.vision.graphic.StandardBoundsScaler

/**
 * Scaler to scale face bounds. This scaler is needed for receivers that need face objects
 * bounding box in the same coordinate system as [GraphicOverlay].
 *
 * We decided to make a shallow copy here fo landmarks because the way we're using the scaler
 * is in a way that we don't touch the landmarks, hence no importance to its modifications
 * is arrtibuted
 */
class FaceBoundsScaler(overlay: GraphicOverlay) : StandardBoundsScaler<FaceObject>(overlay) {
    override fun FaceObject.clone(sourceSize: Size, boundingBox: RectF?): FaceObject =
        copy(sourceSize = sourceSize, boundingBox = boundingBox)
}

/**
 * Scaler to translate the nose landmark. This scaler is needed for receivers that need face objects
 * landmarks in the same coordinate system as [GraphicOverlay].
 *
 * This translator is only used for liveness face since we need to do some processing on
 * the nose coordinate.
 */
class FaceNoseTranslator(overlay: GraphicOverlay) : StandardBoundsScaler<FaceObject>(overlay) {
    override fun FaceObject.clone(sourceSize: Size, boundingBox: RectF?): FaceObject {
        // her instead of overriding transform to scale the landmark points
        // we take the advantage of the fact that [StandardBoundsScaler] computes
        // all the scales parameter before calling this method, hence we can use those
        // scales to fix landmark points.
        // We do two things at once: 1. copying the upstream object. 2. fix the landmarks points.
        val landmarks = mutableListOf<Landmark>().let { list ->
            for (l in this) {
                list.add(if (l is Nose) Nose(position = PointF(
                    translateX(l.position.x),
                    translateY(l.position.y),
                )) else l)
            }
        }
        return copy(sourceSize = sourceSize, boundingBox = boundingBox, landmarks = landmarks)
    }
}

File: react-native/react-native-clockin-vision/ios/Face/FaceTransformer.swift
//
//  FaceTransformer.swift
//  TOTVSClockInVision
//
//  Created by Jansel Rodriguez on 8/4/20.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation
import TOTVSCameraKit

/**
* Scaler to translate the nose landmark. This scaler is needed for receivers that need face objects
* landmarks in the same coordinate system as [GraphicOverlay].
*
* This translator is only used for liveness face since we need to do some processing on
* the nose coordinate.
*/
public class FaceNoseTranslator : Transformer {
    private weak var cameraView: CameraView? = nil
    
    public init(cameraView: CameraView) {
        self.cameraView = cameraView
    }
    
    public func transform(value: FaceObject, receiver: VisionReceiver<FaceObject>) {
        // we do nothing when we receives a null face
        guard NullFaceObject != value, nil != cameraView else {
            return receiver.send(value: value)
        }
        var face = value
        // we do nothing when we don't have a nose
        guard let nose = face[.nose] else {
            return receiver.send(value: face)
        }
        let normalized = normalize(point: nose.position, sourceSize: face.sourceSize)
        
        face[.nose] = Landmark(name: .nose, position: normalized)
        face.sourceSize = cameraView!.frame.size
        
        receiver.send(value: face)
    }
}


fileprivate extension FaceNoseTranslator {
    // normalize points comming from camera device coordinate system.
    private func normalize(point: CGPoint, sourceSize: CGSize) -> CGPoint {
        // this step is required to create a unit point
        let normalized = CGPoint(x: point.x / sourceSize.width, y: point.y / sourceSize.height)
        return cameraView!.previewPoint(fromCaptureDevicePoint: normalized)
    }
}
